### PR TITLE
Test for reading JDK 8 (classfile format 52) class files.

### PR DESCRIPTION
**Pull request by JamesIry**
This commit includes a test for reading JDK 8 (classfile format 52)
class files, in particular default (aka defender) methods. It uses ASM
to generate an interface with default methods then exercises that
interface from Scala.

Surprisingly no changes are necessary to the Scala code base to support
reading format 52 class files.

Because the test can only run under JDK 8, the JDK version is checked
and the expected output is synthesized for previous versions.

### CODE REVIEW

**Comment by JamesIry** [comment]
review @gkossakowski  

Note: I've run a hacked version of the code that throws an error on JDK < 8 to verify that it's actually testing something. To get JDK visit https://jdk8.java.net/download.html

**Comment by retronym** [comment]
The JDK version checking is done a little differently here: b2c67b328. Might pop this into a utility method in partest.

**Comment by JamesIry** [comment]
Ah, @paulp  did the Java 8 classfile work https://github.com/scala/scala/pull/2456. So let me rephrase, "surprisingly it worked in spite of not having a test for class file reading." So there!

**Comment by JamesIry** [comment]
@retronym good idea. And I'll update https://github.com/scala/scala/blob/4c85fb324dc3ea0eeebe590e39e8084558834f3a/test/files/run/classfile-format-51.scala

File: test/files/run/classfile-format-52.scala [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+  override def show(): Unit = {
+    // redirect err to out, for logging
+    val prevErr = System.err
+    System.setErr(System.out)
+    try {
+      // this test is only valid under JDK 1.8+
+      // cheat a little by using 'ScalaVersion' because it can parse java versions just as well
+      val requiredJavaVersion = ScalaVersion("1.8")
+      val executingJavaVersion = ScalaVersion(System.getProperty("java.specification.version"))
+      if (executingJavaVersion >= requiredJavaVersion) {
```
Comment (soc) [inline test/files/run/classfile-format-52.scala]: `isJavaAtLeast("1.8")`?

**Comment by retronym** [comment]
LGTM. Seeing as it's just tests I'm going to merge, but I'd encourage @gkossakowski to still take a look at this after the fact.

File: src/partest/scala/tools/partest/DirectTest.scala [hunk 2]
```diff
@@ -101,4 +102,30 @@ abstract class DirectTest extends App {
...
     if (isDebug) Console.err println msg
   }
+
+  /**
+   * Run a test only if the current java version is at least the version specified.
+   */
+  def testUnderJavaAtLeast[A](version: String)(yesRun: =>A) = new TestUnderJavaAtLeast(version, { yesRun })
+
+  class TestUnderJavaAtLeast[A](version: String, yesRun: => A) {
+    val javaVersion = System.getProperty("java.specification.version")
```
Comment (gkossakowski) [inline src/partest/scala/tools/partest/DirectTest.scala:112]: Interesting. I didn't know there's property like that.

Comment (JamesIry) [inline src/partest/scala/tools/partest/DirectTest.scala:112]: Yeah, much easier to parse than the full version info.

**Comment by gkossakowski** [comment]
LGTM.

Sorry for the delay, I didn't realize it was sitting in my queue. I'll pay more attention to it now.

File: src/partest/scala/tools/partest/DirectTest.scala (hunk 2 above)
Comment (paulp) [inline src/partest/scala/tools/partest/DirectTest.scala:112]: What the... that property needs a better PR agent!
